### PR TITLE
Do not use the system's tmp dir

### PR DESCRIPTION
**Pull request by teohhanhui**
(split out from #48)

/cc @issackelly

### CODE REVIEW

**Comment by alexjg** [comment]
I think it might be worth making this a configuration option. I think the reason for _not_ using the system temp directory is that it might not be on the same filesystem as the data directory and thus atomic renames may not work (correct me if I'm wrong there). I think this is a good argument for not _defaulting_ to the system temp directory. However, if - like me - you want to run this on your primary desktop and your system temp directory _is_ on the same filesystem as the data directory then the temp directory has the really nice feature of cleaning itself up on every reboot and so if the sync server crashes and leaves data hanging around it will not persist beyond the next boot.

**Comment by teohhanhui** [comment]
Extrapolating from there, actually the user could already override the `TMPDIR` env var (or `TMP` / `TEMP` on Windows). So I think there's no need for a config.

I'm closing this.

**Comment by issackelly** [comment]
Alex, you're right about the need, and I don't really care that it's a config option or not.
Unfortunately Han, I think I still need a config option. We can't assume an overwrite TMP for _the whole application_ just for automerge-repo-rs. Some things do not need an atomic move out of TMP and can benefit from ramfs or something. I think that I could probably work around it but a confg variable that defaults to TMPDIR sounds like an "everbody wins!" option

**Comment by alexjg** [comment]
@issackelly I agree, although I don't care too much about defaulting to `TMPDIR`. I think the data directory itself is a good default (i.e. the least likely to result in bug reports for us 😅 ) but there should be the option to override it to whatever the user wants for people who understand the possible downsides.

**Comment by alexjg** [comment]
I think the only thing remaining is to add some brief documentation to the `with_tmpdir` methods explaining what they do (sadly this will have to be duplicated due to the need for duplicate implementations). The most important thing to explain is that the user needs to ensure that the tempdir is on the same filesystem as the data dir.

**Comment by teohhanhui** [comment]
Docs added.

File: src/tokio/fs_storage.rs [hunk 1]
```diff
@@ -21,11 +25,34 @@ impl FsStorage {
...
+    /// The default is to use the root directory passed to [`FsStorage::open`].
+    ///
+    /// The tmpdir used must be on the same mount point as the root directory,
+    /// otherwise the storage will throw an error on writing data.
+    ///
+    /// # Errors
+    ///
+    /// This will attempt to create the tmpdir directory and throw an error if
+    /// it does not exist.
+    pub fn with_tmpdir<P: AsRef<Path>>(self, tmpdir: P) -> Option<Result<Self, std::io::Error>> {
```
Comment (teohhanhui) [inline src/tokio/fs_storage.rs:45]: Is this return type intuitive? `None` is only returned when we can't take ownership of the inner `FsStore`...

**Pull request merged**